### PR TITLE
Support multi-GPU training with DDP from PyTorch.

### PR DESCRIPTION
**Pull request by csukuangfj**
The training time is approximately halved using two GPUs.

But it got stuck at the line
```python
cleanup_dist()
```
when the training was about to finish. So only checkpoints up to `epoch-8.pt` are available.

The decoding results with model averaging using the last 5 checkpoints are
```
2021-04-02 09:31:15,951 INFO [common.py:356] [test-clean] %WER 7.11% [3736 / 52576, 487 ins, 314 del, 2935 sub ]
2021-04-02 09:33:42,194 INFO [common.py:356] [test-other] %WER 18.50% [9686 / 52343, 1080 ins, 889 del, 7717 sub ]
```

The DDP training seems working.

---

After removing `cleanup_dist()` and testing it with a small amount of data, the training process is able to finish in a correct way.
The training process is started again and will run for 10 epochs. I will report the result here after it finishes.

---

Tensorboard logs: 
  - first try (2 GPUs): https://tensorboard.dev/experiment/fJrc3otkQ2WnXHP6pY0nrw/
  - second try (2 GPUs): https://tensorboard.dev/experiment/xY215fYATxyj8d78c5DlLw
  - third try (2 GPUs): https://tensorboard.dev/experiment/LdD8mokDTaWACV8Mbvh5CQ/
  - fourth try (3 GPUs): https://tensorboard.dev/experiment/3YrmNs2dSL6c5lELeg23Fg/

### CODE REVIEW

**Comment by danpovey** [comment]
Cool!

**Comment by pzelasko** [comment]
I think I have an idea what could be causing the hanging. It seems synchronization with NCCL requires using torch.distributed.barrier() before syncing tensors. I have it on my list to check it but if you are working on it atm maybe you want to try.

**Comment by csukuangfj** [comment]
> . It seems synchronization with NCCL requires using torch.distributed.barrier() before syncing tensors.

Thanks, will try it. 

---

I've set num_epochs to 11 so that even if it hangs, we can still get `epoch-9.pt`. It indeed hung. Will try `torch.distributed.barrier()` before exiting the process.

---

Here are the WERs:

(1) averaging with checkpoints from the last 5 epochs
```
2021-04-02 20:50:11,390 INFO [common.py:356] [test-clean] %WER 7.00% [3678 / 52576, 493 ins, 290 del, 2895 sub ]
2021-04-02 20:52:20,198 INFO [common.py:356] [test-other] %WER 18.15% [9501 / 52343, 1058 ins, 942 del, 7501 sub ]
```

It is slightly worse than the results recorded in RESULTS.md:  `6.90` vs `7.00` and `17.89` vs `18.15`. 

Will report WERs per epoch later.

As for the training time, we can use the time points when learning rate data is recorded in the tensorboard log to infer the training time per epoch
- #143 https://tensorboard.dev/experiment/h3xiWY0oQ4WGd2dRgG8NWw/
   - About 52 minutes per epoch
  
- this pull-request (with 2 GPUs):  https://tensorboard.dev/experiment/LdD8mokDTaWACV8Mbvh5CQ
   - About 26 minutes per epoch
  
You can see that the training time is halved with 2 GPUs and the resulting WERs are comparable to that trained with single GPU.

**Comment by pzelasko** [comment]
Cool. Probably we need to tune the learning rate to account for larger effective batch sizes.

As to distributed: I was recently using Lhotse in another project where the training loop is implemented with Pytorch Lightning, and multi-GPU training did never hang on this one. I did an initial scan of what they're doing in Lightning, and [that's how I found about the barrier thing](https://github.com/PyTorchLightning/pytorch-lightning/blob/217c12a4e7251681d3840b70db54e1f733b8d67c/pytorch_lightning/utilities/distributed.py#L120). Maybe there is other useful stuff to be gained from there.

**Comment by csukuangfj** [comment]
> Maybe there is other useful stuff to be gained from there.

Thanks for the information. Will take a look into it.

File: egs/librispeech/asr/simple_v1/mmi_att_transformer_train.py [hunk 1]
```diff
@@ -231,14 +234,14 @@ def train_one_epoch(dataloader: torch.utils.data.DataLoader,
                     tb_writer: SummaryWriter,
                     num_epochs: int,
                     global_batch_idx_train: int):
-    """One epoch training and validation.
+    '''One epoch training and validation.
```
Comment (pzelasko) [inline egs/librispeech/asr/simple_v1/mmi_att_transformer_train.py:237]: I think it's a bad pattern to do some many formatting changes in a PR like this; it's very distracting from what's the actual contribution.

File: egs/librispeech/asr/simple_v1/mmi_att_transformer_train.py [hunk 2]
```diff
@@ -585,7 +608,7 @@ def main():
         cuts_dev,
         max_duration=max_duration,
```
Comment (pzelasko) [inline egs/librispeech/asr/simple_v1/mmi_att_transformer_train.py:609]: You have two options regarding the validation dataset:
1) distribute it to all nodes and synchronize dev loss
2) give 100% to all nodes and forego synchronization

By default, Lhotse's samplers are auto-distributing themselves when they detect distributed training. That means currently you are reporting the dev loss on a partial dev set. You need to either sync the loss across nodes or explicitly set rank=0 and world_size=1 to the dev sampler (to disable the auto-distribution mechanism).

It might be a good idea to explicitly set the rank and world size in train samplers too?

Also I'm not sure if this is a good (as in: least surprising) behaviour by Lhotse classes, I'm open to feedback.

**Review by pzelasko — COMMENTED** [review]
I noticed that you are not synchronizing any of the metrics that are being tracked and reported: especially loss, dev loss and num frames seem like important candidates for accurate reporting.

File: egs/librispeech/asr/simple_v1/mmi_att_transformer_train.py (hunk 1 above)
Comment (csukuangfj) [reply egs/librispeech/asr/simple_v1/mmi_att_transformer_train.py]: Ok, reverted.

**Comment by csukuangfj** [comment]
There are two issues for this pull-request:

- You may have to kill the processes manually to free the GPU memory
- The valid objf is for a single node only.

Otherwise, it is ready to merge.

**Comment by danpovey** [comment]
I mentioned somewhere else that the reason for the hang seems to be that the  sampler seems to give a different number of minibatches in different processes.  Merging anyway.
